
library(dynamicTreeCut)
library(fastcluster)
library(WGCNA)
library(dplyr)
library(ggplot2)
# library(DESeq2)
# library(pheatmap)
# library(RColorBrewer)
# library(corrplot)
# library(igraph)
# library(networkD3)
# library(tibble)
# library(reshape2)
# library(clusterProfiler)
# å¯ç”¨å¤šçº¿ç¨‹ï¼ˆå¦‚æœæ”¯æŒï¼‰
enableWGCNAThreads()

source("~/Desktop/02.Coding/RNA-seq/scripts/functions.R")

dds <- readRDS("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/dds.rds")

# datExpr <- counts(dds, normalized = TRUE)
# chickens <- sample_info %>% as.data.frame() %>% filter(Species == "Chicken") %>% pull(SampleID) 
# ducks <- sample_info %>% as.data.frame() %>% filter(Species == "Duck") %>% pull(SampleID) 


# ===============================
# åœ¨ WGCNA ç½‘ç»œåˆ†æä¹‹å‰ï¼Œéœ€è¦å…ˆè¿‡æ»¤æ‰ä½æ–¹å·®çš„åŸºå› ï¼Œè¿™äº›åŸºå› å¾€å¾€æ²¡æœ‰ç‰¹å®šçš„ç”Ÿç‰©å­¦åŠŸèƒ½ï¼Œè¿˜ä¼šå¢åŠ åˆ†æå™ªéŸ³
# ===============================

datExpr <- prepare_data_for_wgcna(dds = dds, quantile = 0.75, min_count = 5)
datExpr.chicken <- prepare_data_for_wgcna(dds = dds[, dds$Species == "Chicken"], quantile = 0.75, min_count = 5)
datExpr.duck <- prepare_data_for_wgcna(dds = dds[, dds$Species == "Duck"], quantile = 0.75, min_count = 5)


# ===============================
# é€šè¿‡ç»˜åˆ¶æ ·æœ¬æ ‘ç­›é€‰æ ·æœ¬
# ===============================

sampleTree <- hclust(dist(t(datExpr)), method = "average")
plot(sampleTree)

removesample <- c("MA-C-4", "BR-C-1")
datExpr <- datExpr[,!colnames(datExpr) %in% removesample] 
datExpr.chicken <- datExpr.chicken[, !colnames(datExpr.chicken) %in% "BR-C-1"]
datExpr.duck <- datExpr.duck[, !colnames(datExpr.duck) %in% "MA-C-4"]

### æŸ¥çœ‹ä¸‰ä¸ªæ•°æ®é›†å…±äº«çš„åŸºå› 
VennDiagram::venn.diagram(
  list(
    Combine = rownames(datExpr),
    Chicken = rownames(datExpr.chicken),
    Duck = rownames(datExpr.duck)
  ), filename = "~/Downloads/test.tiff",
  fill = c( "green", "gold", "darkorchid1")
)

VennDiagram::venn.diagram(
  list(
    Chicken = rownames(datExpr.chicken),
    Duck = rownames(datExpr.duck)
  ), filename = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/è´¨æ§ååŸºå› æ•°é‡.tiff",
  fill = c( "green", "gold")
)


# ===============================
# é€‰æ‹©åˆé€‚çš„è½¯é˜ˆå€¼
# ===============================

# æœ‰ä¸åˆé€‚çš„æ ·æœ¬æ—¶ï¼Œæ²¡æœ‰è‡ªåŠ¨é€‰åˆ°è½¯é˜ˆå€¼ï¼Œåˆ é™¤ä¸åˆé€‚çš„é˜ˆå€¼åå°±å¾—åˆ°äº†åˆé€‚çš„è½¯é˜ˆå€¼

softPower <- auto_select_soft_Power(datExpr = t(datExpr))
softPower.chicken <- auto_select_soft_Power(datExpr = t(datExpr.chicken))
softPower.duck <- auto_select_soft_Power(datExpr = t(datExpr.duck))

softPower.chicken <- 10  #è‡ªåŠ¨æ£€æµ‹æ—¶é€‰æ‹©14ï¼Œæœ‰ç‚¹è¿‡é«˜äº†ï¼Œå› æ­¤ç¨å¾®é™ä½ä¸€ç‚¹

# è½¯é˜ˆå€¼å·®å¼‚ â‰¤ 1   â†’  ç»Ÿä¸€ä½¿ç”¨åŒä¸€ä¸ªè½¯é˜ˆå€¼ âœ…
# è½¯é˜ˆå€¼å·®å¼‚ 2-3   â†’  ä½¿ç”¨å¹³å‡å€¼ä½œä¸ºæŠ˜ä¸­ âš–ï¸  
# è½¯é˜ˆå€¼å·®å¼‚ â‰¥ 4   â†’  å„è‡ªä½¿ç”¨æœ€ä¼˜è½¯é˜ˆå€¼ ğŸ”¬

# ===============================
# æ„å»ºç½‘ç»œï¼Œå¹¶ç»˜åˆ¶æ¨¡å—å›¾
# ã€æ³¨æ„ã€‘WGCNAç½‘ç»œæ„å»ºè¿‡ç¨‹å’ŒDESeq2åŒ…å­˜åœ¨å†²çªï¼ŒåŠ è½½åä¼šå¯¼è‡´ç¨‹åºè¿è¡Œé”™è¯¯
# ===============================

net.combine <- net_and_plot(datExpr = datExpr, softPower = softPower, labeltext = "Combine")
net.chicken <- net_and_plot(datExpr = datExpr.chicken, softPower = softPower.chicken, labeltext = "Chicken",
                            pdffile = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Chicken.æ¨¡å—.pdf")
net.duck <- net_and_plot(datExpr = datExpr.duck, softPower = softPower.duck, labeltext = "Duck",
                         pdffile = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Duck.æ¨¡å—.pdf" )

# ===============================
# æ¨¡å—çš„ç›¸å…³æ€§
# ===============================

MEs.chicken <- moduleEigengenes(t(datExpr.chicken), colors = labels2colors(net.chicken$colors))$eigengenes %>% orderMEs()
colnames(MEs.chicken) <- colnames(MEs.chicken) %>% gsub("ME","", .)
pdf("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Chicken.æ¨¡å—ç›¸å…³æ€§.pdf", width = 4, height = 4)
pheatmap::pheatmap(mat = cor(MEs.chicken))
dev.off()

MEs.duck <- moduleEigengenes(t(datExpr.duck), colors = labels2colors(net.duck$colors))$eigengenes %>% orderMEs()
colnames(MEs.duck) <- colnames(MEs.duck) %>% gsub("ME","", .)
pdf("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Duck.æ¨¡å—ç›¸å…³æ€§.pdf", width = 4, height = 4)
pheatmap::pheatmap(mat = cor(MEs.duck))
dev.off()

# ===============================
#  Module preservation analysis
# ===============================


multiData <- list(
  Chicken = list(data = t(datExpr.chicken)), # Transpose to make samples columns
  Duck = list(data = t(datExpr.duck)) # Transpose to make samples columns
)

multiColor <- list(
  Chicken = labels2colors(net.chicken$colors) %>% setNames(net.chicken$colors),
  Duck = labels2colors(net.duck$colors) %>% setNames(net.duck$colors)
)

preservation_chicken_vs_duck <- WGCNA::modulePreservation(
  multiData = multiData,
  multiColor = multiColor,
  referenceNetworks = 1,  # å°†ç¬¬ä¸€ä¸ªæ•°æ®é›†ä½œä¸º reference
  nPermutations = 100, 
  randomSeed = 10086,
  parallelCalculation = TRUE,
  verbose = 3
)


preservation_duck_vs_chicken <- WGCNA::modulePreservation(
  multiData = multiData,
  multiColor = multiColor,
  referenceNetworks = 2,  # å°†ç¬¬ä¸€ä¸ªæ•°æ®é›†ä½œä¸º reference
  nPermutations = 200, 
  randomSeed = 10086,
  parallelCalculation = TRUE,
  verbose = 3
)



preservation_stats <- preservation_chicken_vs_duck$preservation$Z$ref.Chicken$inColumnsAlsoPresentIn.Duck %>%
  mutate(Module = rownames(.)) %>%
  dplyr::filter(Module != "gold") 

p.pre.chicken <- ggplot(data = preservation_stats, aes(x = Module, y = Zsummary.pres, fill = Module))  + geom_bar(stat = "identity") +
  scale_fill_manual(values = preservation_stats$Module) + ggtitle("Chicken as ref")
ggsave(filename = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Chicken.ä¿å®ˆæ€§åˆ†æ.pdf", plot = p.pre.chicken, width = 6, height = 4)

# turquoise å’Œ brown æ¨¡å—çš„å¤ç°æ€§å¾ˆé«˜ï¼ŒæŸ¥çœ‹ä¸€ä¸‹æ˜¯å¦æ˜¯åŠŸèƒ½ä¿å®ˆçš„æ¨¡å—


preservation_stats.duck <- preservation_duck_vs_chicken$preservation$Z$ref.Duck$inColumnsAlsoPresentIn.Chicken %>%
  mutate(Module = rownames(.)) %>%
  dplyr::filter(Module != "gold") 

p.pre.duck <- ggplot(data = preservation_stats.duck, aes(x = Module, y = Zsummary.pres, fill = Module))  + geom_bar(stat = "identity") +
  scale_fill_manual(values = preservation_stats.duck$Module) + ggtitle("Duck as ref")
ggsave(filename = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Duck.ä¿å®ˆæ€§åˆ†æ.pdf", plot = p.pre.duck, width = 6, height = 4)


### ç”¨åˆ—è¡¨å‚¨å­˜æ¨¡å—åŸºå› 

genes_in_module.chicken <- split(net.chicken$colors %>% names, f = labels2colors(net.chicken$colors))
genes_in_module.duck <- split(net.duck$colors %>% names, f = labels2colors(net.duck$colors))


# ===============================
#  GO å’Œ KEGG æ³¨é‡Š
# ===============================

# å„æ¨¡å— GO æ³¨é‡Š
enrichment_GO.chicken <- list()
enrichment_KEGG.chicken <- list()
gtf.chicken <- "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/data/GCF_016699485.2_bGalGal1.mat.broiler.GRCg7b_genomic.gtf"
emapperannotations.chicken <- "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/data/chicken.emapper.annotations"

for (module in genes_in_module.chicken %>% names()) {
  
  enrichment_GO.chicken[[module]] <- GO_enrichment(
    diffgenes = extract_gene_from_gene_vs_gene_pair(genes_in_module.chicken[[module]], side = 1), 
    backgroundgenes = extract_gene_from_gene_vs_gene_pair(SummarizedExperiment::assay(dds) %>% rownames(), side = 1),
    GTF = gtf.chicken, emapperannotations = emapperannotations.chicken
  )
  
  enrichment_KEGG.chicken[[module]] <- KEGG_enrichment(
    diffgenes = extract_gene_from_gene_vs_gene_pair(genes_in_module.chicken[[module]], side = 1), 
    backgroundgenes = extract_gene_from_gene_vs_gene_pair(SummarizedExperiment::assay(dds) %>% rownames(), side = 1),
    GTF = gtf.chicken, emapperannotations = emapperannotations.chicken
  )
  
}

# å„æ¨¡å— KEGG æ³¨é‡Š
enrichment_GO.duck <- list()
enrichment_KEGG.duck <- list()
gtf.duck <- "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/data/GCF_047663525.1_IASCAAS_PekinDuck_T2T_genomic.gtf"
emapperannotations.duck <- "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/data/duck.emapper.annotations"
for (module in genes_in_module.duck %>% names()) {
  
  cat("æ¨¡å— ", module, "\n")
  
  enrichment_GO.duck[[module]] <- GO_enrichment(
    diffgenes = extract_gene_from_gene_vs_gene_pair(genes_in_module.duck[[module]], side = 2), 
    backgroundgenes = extract_gene_from_gene_vs_gene_pair(SummarizedExperiment::assay(dds) %>% rownames(), side = 2),
    GTF = gtf.duck, emapperannotations = emapperannotations.duck
  )
  
  enrichment_KEGG.duck[[module]] <- KEGG_enrichment(
    diffgenes = extract_gene_from_gene_vs_gene_pair(genes_in_module.duck[[module]], side = 2), 
    backgroundgenes = extract_gene_from_gene_vs_gene_pair(SummarizedExperiment::assay(dds) %>% rownames(), side = 2),
    GTF = gtf.duck, emapperannotations = emapperannotations.duck
  ) 
  
}


### ç»˜åˆ¶ç‚¹å›¾

for (module in genes_in_module.chicken %>% names()) {
  cat("æ¨¡å— ", module, "\n")
  p.GO <- anno_plot(object = enrichment_GO.chicken[[module]], title = paste0("GO: Chicken module ", module))
  p.KEGG <- anno_plot(object = enrichment_KEGG.chicken[[module]], title = paste0("KEGG: Chicken module ", module))
  p <- ggpubr::ggarrange(p.GO, p.KEGG, labels = c("A", "B"), ncol = 2)
  ggsave(filename = paste0("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Chicken.",module,".anno.pdf" ), plot = p, width = 14, height = 6)
}

for (module in genes_in_module.duck %>% names()) {
  cat("æ¨¡å— ",module, "\n")
  p.GO <- anno_plot(object = enrichment_GO.duck[[module]], title = paste0("GO: Duck module ", module))
  p.KEGG <- anno_plot(object = enrichment_KEGG.duck[[module]], title = paste0("KEGG: Duck module ", module))
  p <- ggpubr::ggarrange(p.GO, p.KEGG, labels = c("A", "B"), ncol = 2)
  ggsave(filename = paste0("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Duck.",module,".anno.pdf" ), plot = p, width = 14, height = 6)
}


# ==============================================================================
#  Expression Profiles
# ==============================================================================

for (module in genes_in_module.chicken %>% names()) {
  
  submod_df <- datExpr.chicken[genes_in_module.chicken[[module]],] %>% 
    as.data.frame() %>% 
    mutate(gene_id = rownames(.)) %>%
    reshape2::melt() 
  
  p <- ggplot(submod_df, aes(x=variable, y=value, group = gene_id)) +
    geom_line(alpha = 0.3) +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 90)
    ) + 
    labs(x = "treatment",
         y = "normalized expression") +
    ggtitle(paste0(module, " module expression profiles in Chicken"))
  
  filename <- paste0("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Chicken.",module, ".expressionprofiles.pdf" )
  ggsave(filename = filename,  width = 14, height = 4)
  
}


for (module in genes_in_module.duck %>% names()) {
  
  submod_df <- datExpr.duck[genes_in_module.duck[[module]],] %>% 
    as.data.frame() %>% 
    mutate(gene_id = rownames(.)) %>%
    reshape2::melt() 
  
  p <- ggplot(submod_df, aes(x=variable, y=value, group = gene_id)) +
    geom_line(alpha = 0.3) +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 90)
    ) + 
    labs(x = "treatment",
         y = "normalized expression") +
    ggtitle(paste0(module, " module expression profiles in Duck"))
  
  filename <- paste0("~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/Duck.",module, ".expressionprofiles.pdf" )
  ggsave(filename = filename, plot = p,  width = 14, height = 4)
  
}


VennDiagram::venn.diagram(
  list(
    Chicken.Brown = genes_in_module.chicken$brown,
    Duck.Red = genes_in_module.duck$red,
    Duck.Pink = genes_in_module.duck$pink
  ), filename = "~/Downloads/test.tiff",
  fill = c( "brown", "red", "pink")
)



submod_df <- DESeq2::getVarianceStabilizedData(dds)[genes_in_module.chicken$green, colnames(datExpr.duck)]  %>% 
  as.data.frame() %>% 
  mutate(gene_id = rownames(.)) %>%
  reshape2::melt() 

p <- ggplot(submod_df, aes(x=variable, y=value, group = gene_id)) +
  geom_line(alpha = 0.3) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90)
  ) + 
  labs(x = "treatment",
       y = "normalized expression") +
  ggtitle("é¸¡ç»†èƒå‘¨æœŸåŸºå› åœ¨é¸­ä¸­çš„è¡¨è¾¾æƒ…å†µ")

ggsave(filename = "~/Desktop/04.æ¹˜æ¹–å®éªŒå®¤/å®¶ç¦½ç—…æ¯’è¯¾é¢˜/RNA-seq/WGCAN/é¸¡ç»†èƒå‘¨æœŸåŸºå› åœ¨é¸­ä¸­çš„è¡¨è¾¾æƒ…å†µ.pdf", plot = p,  width = 14, height = 4)




# ==============================================================================
#  æŸ¥çœ‹ä¸€ä¸‹é¸¡Brown æ¨¡å—ç‰¹æœ‰çš„åŸºå› åœ¨é¸­ä¸­æ˜¯ä»€ä¹ˆæƒ…å†µ
# ==============================================================================


ChickenUniueImmuneGenes <- setdiff( setdiff(genes_in_module.chicken$brown, genes_in_module.duck$red), genes_in_module.duck$pink)

ChickenUniueImmuneGenes

# æŠŠé¸¡å…ç–«æ¨¡å—ç›¸å¯¹é¸­ç‰¹æœ‰çš„åŸºå› æ‰¾å‡ºæ¥ï¼Œå¹¶è¿›è¡ŒåŠŸèƒ½æ³¨é‡Šï¼ŒæŸ¥çœ‹è¿™äº›åŸºå› æ˜¯ä»€ä¹ˆåŠŸèƒ½
ChickenUniueImmuneGenes_InDuck <- list()
for (module in genes_in_module.duck %>% names()) {
  ChickenUniueImmuneGenes_InDuck[[module]]  <- genes_in_module.duck[[module]][genes_in_module.duck[[module]] %in% ChickenUniueImmuneGenes]
}

enrichment_GO.test <- GO_enrichment(
  diffgenes = ChickenUniueImmuneGenes_InDuck$brown %>% extract_gene_from_gene_vs_gene_pair(side = 2), 
  backgroundgenes = extract_gene_from_gene_vs_gene_pair(SummarizedExperiment::assay(dds) %>% rownames(), side = 2),
  GTF = gtf.duck, emapperannotations = emapperannotations.duck
)

anno_plot(object = enrichment_GO.test, title = "unique gene anno by duck")

enrichment_GO.test@result -> test

